Fixes #15407 - adding event queue and listener

jlsherrill opened the pull request:
This adds a new queing table and model as well
as a dynflow listening task that processes queue items.
The only current event is the import errata task, to serialize
errata import so that we can resolve unique constraint
violations once and for all.

jlsherrill commented:
TODO:

*[ ]  write tests
*[ ] validate failure scenarios

jlsherrill commented:
Still have a couple TODO's, but opening for early feedback

ehelms commented on app/models/katello/event_queue.rb:
```diff
@@ -0,0 +1,25 @@
+  class EventQueue < Katello::Model
+    self.table_name = 'katello_event_queue'
+
+    @repository_types = {}
```
What does `repository` imply here? In our world, it can be an overloaded or confusing term.

jlsherrill replied:
apparently my brain checked out when i was creating this class :)  That should be event_types

ehelms commented on app/models/katello/event_queue.rb:
```diff
@@ -0,0 +1,25 @@
+    end
+
+    def self.push_event(event_type, id)
+      self.create!(:event_type => event_type, :object_id => id)
```
What happens when an event_type is specified that is not registered?

jlsherrill replied:
apparently my brain checked out when i was creating this class :)  That should be event types

jlsherrill replied:
I'll add throwing a more specific exception to make it obvious

ehelms replied:
You could add a validator since the underlying object is an active record object.

ehelms commented on app/models/katello/events/import_host_errata.rb:
```diff
@@ -0,0 +1,17 @@
+module Katello
+  module Events
+    class ImportHostErrata < Katello::Model
```
Does this need to be an active record object?

beav commented:
@jlsherrill I'm not 100% familiar with this area, this is an event queue that's kept inside the running instance and does not use amqp at all?

ehelms commented on app/models/katello/event_queue.rb:
```diff
@@ -0,0 +1,25 @@
+module Katello
+  class EventQueue < Katello::Model
```
My object oriented brain feels like this should be an Event with a class that represents the EventQueue and manages the events vs an EventQueue managing instances of event queues.

jlsherrill commented:
@beav correct, this uses a simple db table to de-duplicate work and serialize its insertion.  We potentially could have used the AMQP queue, but this seemed much simpler

beav commented:
@jlsherrill I agree, I think if it's an internal mechanism of Katello, there's no need to expose it via qpid.

ehelms commented on app/lib/actions/katello/monitor_event_queue.rb:
```diff
@@ -0,0 +1,128 @@
+module Actions
```
For me, the triple, or quadrupel embedded classes makes it hard to parse what is going on and what belongs to what in this class (e.g. the `class << self` block).

ehelms commented on app/lib/actions/katello/monitor_event_queue.rb:
```diff
@@ -0,0 +1,128 @@
+      end
+
+      def run(event = nil)
+        Rails.logger.error("FOOO #{event.to_s}")
```
`FOOOO`

jlsherrill replied:
BAAAAAAAAAARRRRRRRRRRR

ehelms commented on app/models/katello/events/import_host_errata.rb:
```diff
@@ -0,0 +1,17 @@
+      end
+
+      def run
+        if @host
```
Could do a `return unless @host` here

ehelms commented on app/models/katello/events/import_host_errata.rb:
```diff
@@ -0,0 +1,17 @@
+      EVENT_TYPE = 'import_host_errata'.freeze
+
+      def initialize(object_id)
+        @host = ::Host.find(object_id)
```
I lean towards handling an invalid host ID properly here so as to produce a better error than the full traceback that would normally occur?

ehelms commented on app/models/katello/event_queue.rb:
```diff
@@ -0,0 +1,25 @@
+
+    def self.pop_event
+      first = Katello::EventQueue.order(:created_at => 'asc').limit(1)
+      events = Katello::EventQueue.where(:object_id => first.pluck(:object_id), :event_type => first.pluck(:event_type)).destroy_all
```
Since this is doing a destroy when it pops them off, is this safe against listener failure and restarts?

beav commented on db/migrate/20160613150922_add_event_queue.rb:
```diff
@@ -0,0 +1,11 @@
+class AddEventQueue < ActiveRecord::Migration
+  def change
+    create_table :katello_event_queue, :force => true do |t|
```
for my understanding of how this works, this table should normally be empty?

jlsherrill replied:
correct.  event gets picked up (marked as in progress), event is executed, event is then deleted form the table.

ehelms commented on app/lib/actions/katello/event_queue/monitor.rb:
```diff
@@ -0,0 +1,107 @@
+          ensure
+            ::Katello::EventQueue.clear_events(event.event_type, event.object_id, event.created_at)
+          end
+          suspend
```
Should we put this in the `begin` block since it would only get run if that is successful? Here it sorta looks like it should run always but due to the rescue and ensure that isn't the case.

ehelms commented on app/lib/actions/katello/event_queue/poller_thread.rb:
```diff
@@ -0,0 +1,44 @@
+                  suspended_action.notify_queue_item(event.event_type, event.object_id, event.created_at) if event
+                end
+
+                sleep 2
```
Just curious if there is any data or precedence for choosing 2?

jlsherrill replied:
hah,  5 seemed too long, 1 seemed too short.  I'm open to something different ;)

beav replied:
I would move this to a constant somewhere, just so its more obvious in case it needs to be tweaked later

ehelms commented on app/services/katello/event_queue.rb:
```diff
@@ -0,0 +1,38 @@
+      Katello::Event.where(:in_progress => true, :object_id => object_id, :event_type => event_type).where('created_at <= ?', on_or_earlier_than).destroy_all
+    end
+
+    def self.peek_event
```
Do I read it correct that this function is setting all events that match the top of queue to in progress? If yes, I would not expect that a function named "peek" to change any date since it implies a viewing rather than modifying type of action.

jlsherrill replied:
suggestions?

ehelms replied:
`pop_event` ? I realize this does not remove it from the queue but it is this queues act of popping the top one off.

![popop](https://cloud.githubusercontent.com/assets/293363/16088301/78e208d8-32f4-11e6-8915-046920371bbc.gif)

jlsherrill replied:
but its not popping it off the top.  to me peek is a better term than pop, but i can see how both are incorrect.  'grab_task' ?

ehelms replied:
`next_event` ?

jlsherrill replied:
Works for me
On Jun 15, 2016 12:51 PM, "Eric D Helms" <notifications@github.com> wrote:

> In app/services/katello/event_queue.rb
> <https://github.com/Katello/katello/pull/6123#discussion_r67200511>:
>
> > @@ -0,0 +1,38 @@
> > +module Katello
> > +  class EventQueue
> > +    @event_types = {}
> > +
> > +    def self.clear_events(event_type, object_id, on_or_earlier_than)
> > +      Katello::Event.where(:in_progress => true, :object_id => object_id, :event_type => event_type).where('created_at <= ?', on_or_earlier_than).destroy_all
> > +    end
> > +
> > +    def self.peek_event
>
> next_event ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Katello/katello/pull/6123/files/83efd9882ca1f8eabd853d6f69f93ed2506eac1a#r67200511>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAYHRcl7pjseJR_xt_G70xIvfC5-w0fvks5qMC2TgaJpZM4I1UO4>
> .
>

jlsherrill commented:
Addressed comments>  Changed the way some of the stuff worked after testing failure scenarios.  It appears that foreman-tasks already does not handle db restarts so that still is not handled after this.  

Filed:  http://projects.theforeman.org/issues/15430

When testing you can run: `Katello::EventQueue.push_event("import_host_errata", 1)`

to throw things onto the queue and see how it handles it.  I still have tests to write & fix

jlsherrill commented:
This should now be ready for final testing/review

beav commented on app/lib/actions/katello/event_queue/poller_thread.rb:
```diff
@@ -0,0 +1,45 @@
+  module Katello
+    module EventQueue
+      class PollerThread
+        SLEEP_INTERVAL = 2
```
![snorlax](http://i.imgur.com/O509E7P.png)

beav commented on app/services/katello/event_queue.rb:
```diff
@@ -0,0 +1,39 @@
+
+    def self.next_event
+      first = Katello::Event.where(:in_progress => false).order(:created_at => 'asc').limit(1)
+      events = Katello::Event.where(:in_progress => false, :object_id => first.pluck(:object_id),
```
When would the events query return more than one result?

jlsherrill replied:
If there are multiple events with the same object_id and event_type yes.  This could happen if for example two pulp applicability generation tasks finish within ~5 seconds of each other (or whatever the dynflow poll length is) and thus the applicability generation dynflow task sticks two on the queue.

beav replied:
Is there an easy way to pause the consumer? I would like to pause during testing and let the queue pile up a bit, then re-enable and confirm it drains

jlsherrill replied:
On a dev environment just:

1) shut down your rails server (this stops dynflow)
2) in a rails console run:
```
 Katello::EventQueue.push_event("import_host_errata", 1)
```

where '1' is the host id.  Run this as many times as you want!
When you're ready to start pulling things off the queue you can just start your rails server.   You can also put invalid host ids on the queue to simulate an error.

A 2nd option would be to just comment out `::Actions::Katello::EventQueue::Monitor.ensure_running(world)
` in engine.rb and restart your rails server.  That'll keep the monitor from running.

beav commented on db/migrate/20160613150922_add_event_queue.rb:
```diff
@@ -0,0 +1,12 @@
+      t.timestamps
+    end
+
+    add_index :katello_events, [:object_id, :event_type, :in_progress], :name => :katello_events_oid_et_ip
```
if the table gets really big (for example, if the queue consumer was dead for awhile), you may need an index on `created_at` to speed up finding the next event

jlsherrill replied:
good point, i'll go ahead and add that

beav commented:
This patch works functionality-wise. For any other PR testers: when you load the rails server, you'll need to hit a page to get the tasks to load up to drain the queue :bear: 

The only issue I saw was that when stopping rails server, the dynflow task did not go to "stopped/success", it stayed as "running". If I stopped and started my server 3x, it appeared as though I had three running consumers at once.

jlsherrill commented:
@beav the shutting down issue should be resolved now

beav commented:
@jlsherrill it works the same as the candlepin listener now in that respect :frog: 

APJ from me!

jlsherrill commented:
Flipping to 'needs re-review' as @ehelms wanted to review it some more

ehelms commented:
ACK - with usability comment

The current implementation doesn't make it easy to understand and know that a task triggered an applicability event that needs to be processed and that said event got processed. I don't want to hold up this change due to its importance but I think that my prior point is worth consideration.

jlsherrill commented:
Agreed, opened http://projects.theforeman.org/issues/15547
